File: patterns/07_command/csharp/Mandatum.cs
// ============================================================
//  COMMAND PATTERN — Imperial Building Commands
//  Run: dotnet script Mandatum.cs
// ============================================================
using System; using System.Collections.Generic; using System.Linq;

class RomeArchitect {
    private List<string> _b = new();
    public string Build(string s)    { _b.Add(s);    return $"🏛  {s} BUILT! [{string.Join(", ", _b)}]"; }
    public string Demolish(string s) { _b.Remove(s); return $"🔨 {s} DEMOLISHED! [{string.Join(", ", _b)}]"; }
}
interface IImperialCommand { string Execute(); string Undo(); }
class BuildCmd : IImperialCommand {
    private readonly RomeArchitect _a; private readonly string _b;
    public BuildCmd(RomeArchitect a, string b) { _a = a; _b = b; }
    public string Execute() => _a.Build(_b);
    public string Undo()    => _a.Demolish(_b);
}
class ImperialPalace {
    private Stack<IImperialCommand> _h = new(), _u = new();
    public string Issue(IImperialCommand c)   { var r = c.Execute(); _h.Push(c); _u.Clear(); return r; }
    public string UndoLast() { if (_h.Count==0) return "Nothing!"; var c=_h.Pop(); _u.Push(c); return c.Undo(); }
    public string RedoLast() { if (_u.Count==0) return "Nothing!"; var c=_u.Pop(); _h.Push(c); return c.Execute(); }
}

Console.WriteLine("╔═══════════════════════════════════════════════╗");
Console.WriteLine("║   COMMAND PATTERN — C#                        ║");
Console.WriteLine("╚═══════════════════════════════════════════════╝\n");
var arch = new RomeArchitect(); var palace = new ImperialPalace();
Console.WriteLine("── ISSUING ──────────────────────────────────────");
Console.WriteLine(palace.Issue(new BuildCmd(arch, "Colosseum")));
Console.WriteLine(palace.Issue(new BuildCmd(arch, "Aqua Claudia")));
Console.WriteLine(palace.Issue(new BuildCmd(arch, "Temple of Jupiter")));
Console.WriteLine("\n── UNDO ─────────────────────────────────────────");
Console.WriteLine(palace.UndoLast()); Console.WriteLine(palace.UndoLast());
Console.WriteLine("\n── REDO ─────────────────────────────────────────");
Console.WriteLine(palace.RedoLast());
Console.WriteLine("\n\"Mandatum datum, mandatum executum!\"");

File: patterns/07_command/php/mandatum.php
<?php
/**
 * COMMAND PATTERN — Imperial Building Commands
 * Run: php mandatum.php
 */
declare(strict_types=1);

class RomeArchitect {
    private array $buildings = [];
    public function build(string $s): string {
        $this->buildings[] = $s;
        return "🏛  {$s} BUILT! [" . implode(", ", $this->buildings) . "]";
    }
    public function demolish(string $s): string {
        $this->buildings = array_values(array_filter($this->buildings, fn($b) => $b !== $s));
        return "🔨 {$s} DEMOLISHED! [" . implode(", ", $this->buildings) . "]";
    }
}
interface ImperialCommand {
    public function execute(): string;
    public function undo():    string;
}
class BuildCmd implements ImperialCommand {
    public function __construct(private RomeArchitect $arch, private string $building) {}
    public function execute(): string { return $this->arch->build($this->building); }
    public function undo():    string { return $this->arch->demolish($this->building); }
}
class ImperialPalace {
    private array $history = [];
    private array $undone  = [];
    public function issue(ImperialCommand $cmd): string {
        $r = $cmd->execute(); $this->history[] = $cmd; $this->undone = []; return $r;
    }
    public function undoLast(): string {
        if (!$this->history) return "Nothing to undo!";
        $cmd = array_pop($this->history); $this->undone[] = $cmd; return $cmd->undo();
    }
    public function redoLast(): string {
        if (!$this->undone) return "Nothing to redo!";
        $cmd = array_pop($this->undone); $this->history[] = $cmd; return $cmd->execute();
    }
}

echo "╔═══════════════════════════════════════════════╗\n";
echo "║   COMMAND PATTERN — PHP                       ║\n";
echo "║   Imperial Building Commands                  ║\n";
echo "╚═══════════════════════════════════════════════╝\n\n";
$arch = new RomeArchitect(); $palace = new ImperialPalace();
echo "── ISSUING COMMANDS ────────────────────────────\n";
echo $palace->issue(new BuildCmd($arch, "Colosseum"))         . "\n";
echo $palace->issue(new BuildCmd($arch, "Aqua Claudia"))      . "\n";
echo $palace->issue(new BuildCmd($arch, "Temple of Jupiter")) . "\n";
echo "\n── UNDO ────────────────────────────────────────\n";
echo $palace->undoLast() . "\n";
echo $palace->undoLast() . "\n";
echo "\n── REDO ────────────────────────────────────────\n";
echo $palace->redoLast() . "\n";
echo "\n\"Mandatum datum, mandatum executum!\"\n";
